server/runtime/src/main/java/org/infinispan/server/security/realm/CachingSecurityRealm.java

```diff
@@ -0,0 +1,322 @@
+import java.util.Arrays;
+
+import org.wildfly.security.auth.SupportLevel;
+import org.wildfly.security.auth.realm.CacheableSecurityRealm;
+import org.wildfly.security.auth.server.IdentityCredentials;
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.auth.server.SecurityRealm;
+import org.wildfly.security.auth.server.event.RealmEvent;
+import org.wildfly.security.authz.Attributes;
+import org.wildfly.security.authz.AuthorizationIdentity;
+import org.wildfly.security.cache.RealmIdentityCache;
+import org.wildfly.security.credential.Credential;
+import org.wildfly.security.credential.PasswordCredential;
+import org.wildfly.security.evidence.Evidence;
+import org.wildfly.security.evidence.PasswordGuessEvidence;
+import org.wildfly.security.password.Password;
+import org.wildfly.security.password.interfaces.ClearPassword;
+
+/**
+ * <p>A wrapper class that provides caching capabilities for a {@link SecurityRealm} and its identities.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class CachingSecurityRealm implements SecurityRealm {
+
+   private final CacheableSecurityRealm realm;
+   private final RealmIdentityCache cache;
+
+   /**
+    * Creates a new instance.
+    *
+    * @param realm the {@link SecurityRealm} whose {@link RealmIdentity} should be cached.
+    * @param cache the {@link RealmIdentityCache} instance
+    */
+   public CachingSecurityRealm(CacheableSecurityRealm realm, RealmIdentityCache cache) {
+      this.realm = checkNotNullParam("realm", realm);
+      this.cache = checkNotNullParam("cache", cache);
+
+      if (realm instanceof CacheableSecurityRealm) {
+         CacheableSecurityRealm cacheable = (CacheableSecurityRealm) realm;
+         cacheable.registerIdentityChangeListener(this::removeFromCache);
+      } else {
+         throw new IllegalArgumentException(realm.toString());
+      }
+   }
+
+   @Override
+   public RealmIdentity getRealmIdentity(Principal principal) throws RealmUnavailableException {
+      RealmIdentity cached = cache.get(principal);
+
+      if (cached != null) {
+         if(log.isTraceEnabled()) {
+            log.tracef("Returning cached RealmIdentity for '%s'", principal);
+         }
+         return cached;
+      }
+
+      RealmIdentity realmIdentity = getCacheableRealm().getRealmIdentity(principal);
+
+      if (!realmIdentity.exists()) {
+         if(log.isTraceEnabled()) {
+            log.tracef("RealmIdentity for '%s' does not exist, skipping cache.'", principal);
+         }
+         return realmIdentity;
+      }
+
+      RealmIdentity cachedIdentity = new RealmIdentity() {
+         final RealmIdentity identity = realmIdentity;
+
+         AuthorizationIdentity authorizationIdentity = null;
+         Attributes attributes = null;
+         IdentityCredentials credentials = IdentityCredentials.NONE;
+         int verifiedEvidenceHash;
```

Comment:
So it's not enough to cache the `RealmIdentity`, we also have to cache the `AuthorizationIdentity` and `IdentityCredentials`?